drivers/sensor/mpu9250/ak8963.c

```diff
@@ -0,0 +1,348 @@
+	 */
+
+	/* If overflow happens then value is invalid */
+	if ((st2 & AK8963_ST2_OVRFL_BIT) != 0) {
+		LOG_INF("Magnetometer value overflow.");
+		return -EOVERFLOW;
+	}
+
+	int64_t scaled_val = (int64_t)raw_val * (int64_t)scale;
+
+	val->val1 = scaled_val / 1000000;
+	val->val2 = scaled_val % 1000000;
+	return 0;
+}
+
+
+static int ak8963_execute_rw(const struct device *dev, uint8_t reg, bool write)
+{
+	/* Instruct the MPU9250 to access over its external i2c bus
+	 * given device register with given details
+	 */
+	const struct mpu9250_config *cfg = dev->config;
+	uint8_t mode_bit = 0x00;
+	uint8_t status;
+	int ret;
+
+	if (write == false) {
+		mode_bit = I2C_READ_FLAG;
+	}
+
+	/* Set target i2c address */
+	ret = i2c_reg_write_byte_dt(&cfg->i2c,
+				    MPU9250_REG_I2C_SLV4_ADDR,
+				    AK8963_I2C_ADDR | mode_bit);
+	if (ret < 0) {
+		LOG_ERR("Failed to write i2c target slave address.");
+		return ret;
+	}
+
+	/* Set target i2c register */
+	ret = i2c_reg_write_byte_dt(&cfg->i2c,
+				    MPU9250_REG_I2C_SLV4_REG,
+				    reg);
+	if (ret < 0) {
+		LOG_ERR("Failed to write i2c target slave register.");
+		return ret;
+	}
+
+	/* Initiate transfer  */
+	ret = i2c_reg_write_byte_dt(&cfg->i2c,
+				    MPU9250_REG_I2C_SLV4_CTRL,
+				    MPU9250_REG_I2C_SLV4_CTRL_VAL);
+	if (ret < 0) {
+		LOG_ERR("Failed to initiate i2c slave transfer.");
+		return ret;
+	}
+
+	/* Wait for a transfer to be ready */
+	while (!(status & MPU9250_I2C_MST_STS_SLV4_DONE)) {
+		ret = i2c_reg_read_byte_dt(&cfg->i2c,
+					   MPU9250_I2C_MST_STS, &status);
+		if (ret < 0) {
+			LOG_ERR("Waiting for slave failed.");
+			return ret;
+		}
+
+		k_msleep(10);
+	}
+
+	return 0;
+}
+
+static int ak8963_read_reg(const struct device *dev, uint8_t reg, uint8_t *data)
+{
+	const struct mpu9250_config *cfg = dev->config;
+	int ret;
+
+	/* Execute transfer */
+	ret = ak8963_execute_rw(dev, reg, false);
+	if (ret < 0) {
+		LOG_ERR("Failed to prepare transfer.");
+		return ret;
+	}
+
+	/* Read the result */
+	ret = i2c_reg_read_byte_dt(&cfg->i2c,
+				   MPU9250_REG_I2C_SLV4_DI, data);
+	if (ret < 0) {
+		LOG_ERR("Failed to read data from slave.");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int ak8963_write_reg(const struct device *dev, uint8_t reg, uint8_t data)
+{
+	const struct mpu9250_config *cfg = dev->config;
+	int ret;
+
+	/* Set the data to write */
+	ret = i2c_reg_write_byte_dt(&cfg->i2c,
+				    MPU9250_REG_I2C_SLV4_DO, data);
+	if (ret < 0) {
+		LOG_ERR("Failed to write data to slave.");
+		return ret;
+	}
+
+	/* Execute transfer */
+	ret = ak8963_execute_rw(dev, reg, true);
+	if (ret < 0) {
+		LOG_ERR("Failed to transfer write to slave.");
+		return ret;
+	}
+
+	return 0;
+}
+
+
+static int ak8963_set_mode(const struct device *dev, uint8_t mode)
+{
+	int ret;
+
+	ret = ak8963_write_reg(dev, AK8963_REG_CNTL1, mode);
+	if (ret < 0) {
+		LOG_ERR("Failed to set AK8963 mode.");
+		return ret;
+	}
+
+	/* Wait for mode to change */
+	k_msleep(1);
```

Comment:
magic number: needs to be defined, why 1ms?

Reply:
Defined